src/Compilers/Core/Portable/SourceGeneration/SyntaxStore.cs

```diff
@@ -121,6 +136,34 @@ public IStateTable GetSyntaxInputTable(SyntaxInputNode syntaxInputNode, NodeStat
                 return result;
             }
 
+            /// <summary>
+            /// Gets the adjustment to wall clock time that should be applied for a set of input nodes.
+            /// </summary>
+            /// <remarks>
+            /// The syntax store updates all input nodes in parallel the first time an input node is asked to update,
+            /// so that it can share the semantic model between multiple nodes and improve perf. 
+            /// 
+            /// Unfortunately that means that the first generator to request the results of a syntax node will incorrectly 
+            /// have its wall clock time contain the time of all other syntax nodes. And conversely other input nodes will 
+            /// not have the true time taken.
+            /// 
+            /// This method gets the adjustment that should be applied to the wall clock time for a set of input nodes
+            /// so that the correct time is attributed to each.
+            /// </remarks>
+            public TimeSpan GetRuntimeAdjustment(ImmutableArray<SyntaxInputNode> inputNodes)
+            {
+                TimeSpan result = TimeSpan.Zero;
+                foreach (var node in inputNodes)
+                {
+                    // only add if this node ran at all during this pass
+                    if (_syntaxTimes.ContainsKey(node))
```

Comment:
>_syntaxTimes.ContainsKey(node) [](http://example.com/codeflow?start=24&length=30)

Consider using `_syntaxTimes.TryGetValue(out var value)` to avoid two lookups.